### ISSUE TITLE
Rebuttal

### ISSUE THREAD

**Issue by isPANN** [yao-baijian/fem-fpga-placer#5]
# Rebuttal Letter to Reviewers - Submission #108

We thank the reviewers for their constructive feedback. The main concerns are evaluation scale, timing/runtime/HPWL comparisons with Vivado, resource coverage, baseline placers, and the “12.2× scale‑up” claim. Our rebuttal addresses each with new timing results under 100 MHz, clarified Vivado runtime (isolated place_design), VPR comparisons, memory‑scaling data, legalization impact (<2%), and a clear definition of the scale‑up ratio. We also discuss extensibility to other types of instance. We propose to incorporate key results into the paper and relegate technical specifics to supplementary material, which we believe effectively strengthens the claims.

## Response to Reviewer #108A

We thank the reviewer for the positive feedback. The main concerns are: (1) small scale (<3k instances), (2) unfair HPWL comparison with Vivado (timing‑driven), and (3) limited resource handling (logic/I/O only). We address these and the VPR comparison below.

> **[Q1]** Claimed 9.74% HPWL improvement over Vivado is unfair (timing-driven vs. HPWL-only). Please provide post‑route timing (WNS/TNS/Fmax); otherwise comparison is meaningless.

We agree that comparing HPWL against a timing‑driven placer is not apples‑to‑apples. For a fairer assessment, we also ran place‑and‑route under 100 MHz (device: xczu7ev‑ffvc1156‑2‑e), excluding purely combinational (ISCAS85) and unpipelined divider (LU8PEEng) benchmarks. Partial timing results are summarized below.

| Circuit | Vivado WNS (ns) | Vivado Fmax (MHz) | moFEM WNS (ns) | moFEM Fmax (MHz) |
| :---: | :---: | :---: | :---: | :---: |
| s15850 | **6.186** | 262 | 4.399 | 197 |
| s38417 | **5.364** | 215 | 5.178 | 207 |
| ch_intrinsics | **8.905** | - | 7.649 | - |
| bgm | **3.513** | 154 | 3.071 | 143 |
| blob_merge | 4.091 | 111 | **4.585** | 185 |
| sha | **4.354** | 177 | 4.012 | 167 |
| mkDelayWorker32B | 3.411 | 151 | **5.423** | 218 |

The result here shows that in most cases, Vivado has a leading advantage in timing results, aligning the moFEM objective design. Since we would make comparison with quantum approches who only optimize HPWL, we would add timing constraint in our future work.

> **[Q2]** How large a design can a single A100 handle before memory/runtime limits? Scaling analysis needed to gauge industrial gap.

We have conducted a scaling analysis on A100 (40GB) with batch=1. And the largest problem we successfully placed LU64PEEng had N = 15740 instances and M = 19881 (20k) sites, consuming 12.25 GB (predicted 10GB) GPU memory. The runtime was more than 3 hours, which is already prohibitive.

We have analyzed memory usage, and it (single‑precision) comprises: probability p (4NM), field h (4NM), gradients (4NM), Adam m/v (8NM), gradient activations (around 4NM), coupling W (4N²), and distance D (4M²). In A100 memory alone supports up to 40k instances on 40k sites, but runtime becomes prohibitive for LU64PEEng; we thus cap at 15k instances on 20k sites. Scaling to >100k instances would require task partitioning, which we leave as future work and will discuss with a scaling table in the revised manuscript.

> **Q3.** Current QUBO claims heterogeneous support via per‑region penalties, but real resources (DSP/BRAM/clock I/O) have column‑based, multi‑site, and bank/voltage constraints. Can the framework handle these, or need extensions?

We thank the reviewer for raising this important distinction between heterogeneous resource regions and the full set of FPGA legality constraints. Our answer is that these constraints fit the proposed mean-field formulation: they fall into the same optimization template used in the current implementation — legal-domain restriction, smooth occupancy penalties, and small auxiliary categorical marginals.

The current prototype implements the first two mechanisms. Type legality is handled by defining each instance's softmax over its legal site set, and per-region capacity by the smooth hinge occupancy penalty. Column restrictions for DSP/BRAM resources add no new difficulty: once legal columns are known, they are part of the legal site set. The remaining constraints—multi-site macros, bank/voltage compatibility, clock-region budgets—require implementation in the same solver, and we formulate each below.

| Constraint | Current status | How it enters | New variables? |
| :--- | :---: | :--- | :---: |
| Type legality (logic / I/O) | implemented | restrict legal site set | no |
| Per-region capacity | implemented | hinge occupancy penalty | no |
| Column restriction | direct extension | restrict legal columns | no |
| Multi-site / hard macros | extension | footprint occupancy penalty | no |
| Bank / voltage domain | extension | bank-standard auxiliary marginal | yes |
| Clock-region budget | extension | clock-presence auxiliary marginal | yes |

For the three extensions, the added terms are concrete and small. A **hard macro** covering a footprint $\mathrm{fp}_i(j')$ when anchored at $j'$ generalizes the capacity penalty through the expected occupancy $\mathbb{E}[o_j] = \sum_i \sum_{j'} \mathbb{1}[j \in \mathrm{fp}_i(j')]\, p_{ij'}$:

$$ E_{\mathrm{cap}}^{\mathrm{multi}} = \lambda \sum_j \max\!\left(0,\ \mathbb{E}[o_j] - 1\right)^2 $$

A **bank / voltage domain** adds a per-bank standard marginal $q_{bc}$ ($b(j)$ is the bank of site $j$, $\mathrm{incompat}(i,c)\in\{0,1\}$) and a bilinear compatibility penalty:

$$ E_{\mathrm{bank}} = \eta \sum_{i,j,c} \mathrm{incompat}(i,c)\, p_{ij}\, q_{b(j)c} $$

A **clock-region budget** adds a clock-presence marginal $s_{gc}$ for region $g$ (capacity $C$):

$$ E_{\mathrm{budget}} = \rho \sum_g \max\!\left(0,\ \sum_c s_{gc} - C\right)^2 $$

The cross-region wirelength term already appears in the logic–I/O experiments; adding DSP/BRAM regions amounts to new legal site sets and region indices within the same objective. In all three extensions the added term is a smooth function of the placement marginals, so the optimization loop is unchanged — only the per-constraint terms above are introduced.

> **Q4.** Could you provide HPWL comparison against VPR on the benchmarks to contextualize QUBO placement relative to non‑QUBO methods?

We performed this comparison. VPR 9.0 was run with its default placement strategy on the same set of benchmark circuits. Partial results are summarized below:

| Circuit | VPR instance (N) | VPR HPWL | moFEM HPWL |
| :---: | :---: | :---: | :---: | :---: |
| ch_intrinsics | 77 | 1948 | **1655 (15%)** |
| bgm | 552 | 25104 | **23979 (4%)** |
| LU8PEEng | 5041 | 199096 | **173251 (13%)** |
| blob_merge | 882 | 47895 | **46089 (4%)** |
| sha | 340 | 15194 | **12991 (15%)** |
| mkDelayWorker32B | 2167 | 82570 | **57566 (30%)** |
| stereovision0 | 1650 | 53979 | **52161 (3%)** |

Across benchmarks, our placer outperforms VPR, achieving a average 12% wirelength reduction, confirming improved placement quality.

---

## Response to Reviewer #108B

We thank the reviewer for the positive assessment and constructive suggestions.

**Hyperparameter sensitivity** We agree robustness is a common QUBO concern. Our Random Forest predictor auto-tunes penalty coefficients $\lambda$, eliminating manual grid search with modest HPWL cost (Section 3.5). ML-predicted $\lambda$ performs near oracle-tuned, mitigating sensitivity.

**Additional experiment** Per reviewer and #108A/#108C requests, we now include:
- Timing (WNS/Fmax) at 100MHz,
- VPR (SA) HPWL comparison,
- Pre‑ vs post‑legalization HPWL (<2% degradation).

These strengthen moFEM’s placement quality evidence and will be included in the revision.

---

## Response to Reviewer #108C

We thank the reviewer for the constructive feedback and for acknowledging our technical contributions. The main concerns are resource coverage, timing/HPWL comparisons, runtime measurement, the scale‑up claim, virtual I/O details, and legalization impact—all addressed below with additional experiments.

> **Q1. Timing and Vivado mode.** Vivado mode used? Timing results needed (WNS/TNS/Fmax) for both, as Vivado is timing‑driven vs. moFEM HPWL‑only—otherwise 9.74% HPWL gain is inconclusive.

We thank the reviewer for raising this critical point. Initially, when comparing Vivado to moFEM, we did not set a high clock constraint and we believe this would prevent Vivado from trading HPWL for timing closure. For the timing result, as also requested by Reviewer #A (Q1), we have added WNS and Fmax estimation under 100MHz.

> **Q2. Vivado runtime measurement methodology.** Figure 4(b) shows Vivado runtime never below ~8s even for tiny s1488 (22 LUTs), suggesting overhead. Please (a) clarify what the measurement includes, (b) isolate place_design, and (c) report moFEM’s end‑to‑end time for a fair comparison.

We thank the reviewer for pointing that out, and we would like to clarify the experiment details from the following aspects:

(a) Our measurement only times the place_design command after the DCP does not contain any other period. The Vivado placement and timing measure tcl code reads as:

```
set place_start [clock seconds]
place_design
set place_end [clock seconds]
```

(b) The ~8 seconds on tiny circuits (e.g., s1488 with 22 logic instances) is the intrinsic floor runtime of Vivado's placement engine, even if the design is small ([UG904](https://docs.amd.com/r/en-US/ug904-vivado-implementation)).

(c) moFEM's end-to-end cost includes: Preparation (netlist extraction, matrix build, GPU preparation) + iterations (main optimization loop) + Inference (inference of the placement by optimization result) + legalization (fast greedy legalization). And part of the results are given below:

| Circuit | moFEM Preparation (s) | moFEM iterations (s) | moFEM Inference (s) | moFEM Legalization (s) |
| :---: | :---: | :---: | :---: | :---: |
| ch_intrinsics | 0.83 | 0.85 | 0.02 | 0.28 |
| bgm | 1.02 | 32.42 | 1.68 | 1.45 |
| LU8PEEng | 0.77 | 241.49 | 3.08 | 1.28 |
| blob_merge | 0.84 | 12.58 | 0.05 | 1.03 |
| sha | 0.84 | 8.39 | 0.02 | 0.10 |
| mkDelayWorker32B | 0.90 | 8.40 | 0.03 | 1.36 |
| stereovision0 | 0.71 | 22.22 | 0.08 | 0.82 |

The key observation is that for large designs (bgm, LU8PEEng), GPU iterations account for the majority of the total runtime.

> **Q3. Definition of the 12.2× scale-up claim.** The abstract reports "a scale-up of problem size up to 12.2×", but this metric is never precisely defined.

We apologize for the ambiguity appearing in our paper. The "12.2× scale-up" refers to the maximum number of instances (N) that moFEM can handle compared to the baseline SBM-related method's maximum instance number (N_sbm) in the same region. Since in paper, the largest design is LU8PEEng with 2578 logic instances, therefore, N = 2578, and N_sbm is 103, the new scale up = N / N_sbm = 25.02. We would make this clearer in our revised paper.

> **Q4. Extensibility to hard macros.** Only logic and virtual I/O are covered—DSP/BRAM/clock buffers deferred. Can the method extend to >1×1 blocks without major reformulation? A demo on one extra resource type would support the heterogeneous claim.

We thank the reviewer. As also noted in our response to Reviewer #108A (Q3), extending to multi‑site macros, bank/voltage constraints, and clock region budgets is feasible without reformulating the core framework. We can handle them via footprint‑aware occupancy penalties or pairwise “must‑be‑together” constraints—both compatible with the differentiable objective (see Reviewer #108A (Q3) for math formulation). A demonstration on placement with DSP blocks will be included in the revised paper.

> **Q5. Legalization impact.** Report HPWL before/after legalization and the number of moved instances, to clarify how much HPWL advantage survives.

We have added the following table to differentiate pre-legalization HPWL and after-legalization HPWL, and will include it in the final paper:

| Circuit | Pre-legalization HPWL | After-legalization HPWL |
| :---: | :---: | :---: |
| bgm | 105947 | 105944 |
| LU8PEEng | 121375 | 120489 |
| blob_merge | 41538 | 40417 |
| sha | 6201 | 6116 |
| mkDelayWorker32B | 21391 | 21317 |
| stereovision0 | 32103 | 32019 |

Results show that legalization induces negligible HPWL change (<2%), almost fully preserving the wirelength quality, thus validating the placer's initial placement.

---

## Response to Reviewer #108D

We thank the reviewer for the positive feedback. The two main concerns are virtual I/O details and the lack of comparison with traditional placers (VPR/analytical)—both addressed below.

> **Q1.** Can you elaborate more details on the virtual I/O settings? For example, what site type do the virtual I/O ports occupy? Are they constrained to the clock region edge, free to move, or fixed?

Virtual I/O ports are inserted as registers (LUT/flip-flop pairs) placed on general logic sites located along the boundary of each clock region. They are not constrained to fixed I/O pads, and they can be optimized in a constrained region. We have clarified this in the experimental setup.

> **Q2.** Since VTR benchmarks are used, why not use VPR as a placement baseline as well? How does moFEM compare against SA-based or analytical placers?

We appreciate the reviewer for raising this question regarding the VPR benchmark. This is also mentioned by Reviewer #A. We have now included a comparison with VPR (SA-based) in the revised paper (see response to Reviewer #A (Q4)). The result shows that our placer yields a relative geometric wirelength reduction of 12% compared with VPR. For analytical placers such as elfPlace, we are unable to run them for now because most of them were built with a more elaborate HPWL model, which we can not compete with for now.
